Option to set the iothreads

alicefr opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR:
Wasn't possible to declare how many iothreads to use.

After this PR:
This PR introduces the IOThreads field to allow to set the number of IOThreads. Virtqueues cannot  be mapped to multiple iothreads. This [blog](https://developers.redhat.com/articles/2024/09/05/scaling-virtio-blk-disk-io-iothread-virtqueue-mapping#how_iothread_virtqueue_mapping_works) explains in details the configuration and the performance gains.

This option needs to be used with `dedicatedCpuPlacement: true` and  `isolateEmulatorThread: true`. This allocated additional CPUs to the virt-launcher pod corresponding to the set number of iothreads and pin each iothread to a CPU.

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add the iothreads option to specify number of iothreads to be used
```

alicefr commented:
/cc @vladikr @stefanha

This is a draft PR for adding the iothread virtqueues mapping.

alicefr commented:
/test all

stefanha commented:
Looks good to me. I am not familiar with the code, but my understanding is this allows the user to specify the number of IOThreads and then KubeVirt will add an <iothreads> element with N IOThreads to the domain XML (i.e. IOThread Virtqueue Mapping for virtio-blk devices).

The parity `% 2` code confused me, but I didn't dig into what it's doing.

alicefr commented:
> Looks good to me. I am not familiar with the code, but my understanding is this allows the user to specify the number of IOThreads and then KubeVirt will add an element with N IOThreads to the domain XML (i.e. IOThread Virtqueue Mapping for virtio-blk devices).
> 

Correct!

> The parity `% 2` code confused me, but I didn't dig into what it's doing.

This was already in the code, I simply added the additional number of CPUs to keep the calculation correct. It is for preserving an even number of CPUs assigned to the pod.

lyarwood reviewed (commented):
A quick API comment before I look further.

lyarwood commented on staging/src/kubevirt.io/api/core/v1/schema.go:
```diff
@@ -203,6 +203,9 @@ type DomainSpec struct {
 	IOThreadsPolicy *IOThreadsPolicy `json:"ioThreadsPolicy,omitempty"`
+	// IOThreads specifies how many IOThreads are dedicated to the disks.
+	// +optional
+	IOThreads *DiskIOThreads `json:"ioThreads,omitempty"`
```
Yeah fair, it is slightly confusing with `disk.dedicatedIOThread` also present in the API.

alicefr replied:
Well, the IOThreads count is shared between all the disks which has `dedicatedIOThread` field enabled. It could be 3 disks but you want to request 4 IOthreads. Then, the load is spread over the 4 ioThreads.

lyarwood replied:
Ah sorry I misunderstood the meaning of that field, okay that makes sense and I can see why you think `ioThreadsPolicy` is redundant in that case. Definitely a candidate to be deprecated in `v1`.

vladikr commented:
@alicefr I'm wondering if you have considered creating a new policy and specifying the static number of threads there? 

```yaml
spec:
  domain:
    ioThreadsPolicy:
      static:
        count: 4
    cpu:
...
...
```
With this approach, we won't need to add a dedicated API field - I don't see a benefit in having it.
A policy is easier to deprecate or replace with another one.

lyarwood commented:
> @alicefr I'm wondering if you have considered creating a new policy and specifying the static number of threads there?
> 
> ```yaml
> spec:
>   domain:
>     ioThreadsPolicy:
>       static:
>         count: 4
>     cpu:
> ...
> ...
> ```
> 
> With this approach, we won't need to add a dedicated API field - I don't see a benefit in having it. A policy is easier to deprecate or replace with another one.

I don't think we change change the type of a field in a `v1` API like this. As I said above I think deprecating `ioThreadsPolicy` and introducing something more expressive with or without `policy` as an attribute would be cleaner.

vladikr commented:
> > @alicefr I'm wondering if you have considered creating a new policy and specifying the static number of threads there?
> > ```yaml
> > spec:
> >   domain:
> >     ioThreadsPolicy:
> >       static:
> >         count: 4
> >     cpu:
> > ...
> > ...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > With this approach, we won't need to add a dedicated API field - I don't see a benefit in having it. A policy is easier to deprecate or replace with another one.
> 
> I don't think we change change the type of a field in a `v1` API like this. As I said above I think deprecating `ioThreadsPolicy` and introducing something more expressive with or without `policy` as an attribute would be cleaner.

Yes, you're right, my example was not the best.

That said, the currently suggested API isn't structured well enough to accommodate different scenarios that may be needed in the future.
 It's also not entirely clear how a user would determine the right number of IO threads for a VM. 
We’ve discussed offline the idea of external components providing hints for the required thread count, and there was mention that in most cases, 4 to 8 threads would be enough (though I’m still uncertain how one determines the right number).

Given this, perhaps it would make sense to reconsider the IOThreadsPolicyAuto policy and focus on "doing the right thing" by default, until we finalize a more suitable API design.

The scale team's previous guidance was to scale IO threads based on the number of disks and vCPUs. In the case where qemu main thread was isolated, pinning a single iothread to the same pCPU as the emulator thread was considered to be enough.
Now, the approach is shifting and may continue to evolve. Therefore, I believe our API should reflect the user’s intent while remaining flexible enough to accommodate future changes. In many cases, users may not need to configure these low-level details at all.

In general,  I would prefer to abstract this complexity into a policy that lets KubeVirt be smart enough to handle this configuration, rather than requiring users to manage it.

alicefr commented:
> That said, the currently suggested API isn't structured well enough to accommodate different scenarios that may be needed in the future. It's also not entirely clear how a user would determine the right number of IO threads for a VM. We’ve discussed offline the idea of external components providing hints for the required thread count, and there was mention that in most cases, 4 to 8 threads would be enough (though I’m still uncertain how one determines the right number).
> 

The problem remains that KubeVirt doesn't know the best number and this is achieved by measuring the performance on the VM and setting that value from the experiments. So, this can be done only by the user

alicefr commented:
@vladikr I slightly changed the API. Now, the new IOThreads struct has also the policy included there and I added a new policy type which should describe the new type of iothread allocation. We can , then, deprecate the iothreadPolicy in favor of this new field. WDYT?

jeniferh commented:
Is it possible to have an option that does not require `dedicatedCpuPlacement` & `isolateEmulatorThread` ? While that may help performance in some cases, I often see environments that do not enable CPU Manager at the node level which would be a requirement.  To prevent over allocation of threads to physical resources, I could see a case where the number of IOthreads requested adds to the overall pod cpu request, but does not require the strict cpuset pinning.

alicefr commented:
> Is it possible to have an option that does not require `dedicatedCpuPlacement` & `isolateEmulatorThread` ? While that may help performance in some cases, I often see environments that do not enable CPU Manager at the node level which would be a requirement. To prevent over allocation of threads to physical resources, I could see a case where the number of IOthreads requested adds to the overall pod cpu request, but does not require the strict cpuset pinning.

@jeniferh how would you distribute the iothreads? This PR pins each iothreads to the additional pCPU. Additionally, I have put as requested the `isolateEmulatorThread` because otherwise this will interfere either with the vcpu threads or the iothreads.

iholder101 commented:
> It's also not entirely clear how a user would determine the right number of IO threads for a VM. ... In many cases, users may not need to configure these low-level details at all.

> The problem remains that KubeVirt doesn't know the best number and this is achieved by measuring the performance on the VM and setting that value from the experiments. So, this can be done only by the user

Thanks @alicefr for this PR and discussion!

I tend to agree with @vladikr's approach here. If we can't know how to configure this in the most optimized way, it would probably be very difficult for users to do so correctly.

In general, my opinion is that currently Kubevirt tends to give the admin/user a lot more control than he needs. This leads to a complicated API which demands a high expertise to define correctly. Our migration configurations are a good example for this, but there are a lot of other examples.

I would much rather use a declarative policy instead of a concrete number of threads, that will use a simple definition (i.e. always use 4 iothreads) or a heuristic (between 4 and 8 based on number of disks/vCPUs). Moving forward, the burden of optimizing this policy would be on Kubevirt, not on its users. It is possible that in the future we'd face a glass celling w.r.t. optimizing this and will see that users demand more control in order to fine tune, but let's not get ahead of ourselves and start simple. As @vladikr mentioned, going forward we might come up with tools that would be able to fine tune automatically an admin's intervention.

iholder101 commented on pkg/virt-controller/services/template.go:
```diff
@@ -1474,10 +1474,16 @@ func (t *templateService) VMIResourcePredicates(vmi *v1.VirtualMachineInstance,
 
 	metrics.SetVmiLaucherMemoryOverhead(vmi, memoryOverhead)
 	withCPULimits := t.doesVMIRequireAutoCPULimits(vmi)
+	additionalCPUs := uint32(0)
+	if vmi.Spec.Domain.IOThreads != nil &&
+		vmi.Spec.Domain.IOThreads.Policy == v1.IOThreadsPolicyDedicated &&
+		vmi.Spec.Domain.IOThreads.Count != nil {
+		additionalCPUs = *vmi.Spec.Domain.IOThreads.Count
+	}
```
Some technical questions:
1. Is it a requirement that each iothread would need a whole dedicated CPU to run on? In the worst case where all of the iothreads would be busy they would have to compete for CPU time, but is this case common? Should we  for example consider to allocate only 1 additional CPU for all of the iothreads?
2. Is it reasonable to pin the iothreads to the "housekeeping" CPU? Generally this CPU is assumed to not be very busy, and its main job is to isolate "housekeeping" tasks (which aren't CPU bound) from the pCPUs that are used exclusively by vCPUs.

alicefr replied:
> Some technical questions:
> 
> 1. Is it a requirement that each iothread would need a whole dedicated CPU to run on? In the worst case where all of the iothreads would be busy they would have to compete for CPU time, but is this case common? Should we  for example consider to allocate only 1 additional CPU for all of the iothreads?

I've been talking with @stefanha about this, and yes each iothread should be pin to a separate pCPU. Otherwise, just use a single iothread. @stefanha please correct me here if I'm wrong.

> 2. Is it reasonable to pin the iothreads to the "housekeeping" CPU? Generally this CPU is assumed to not be very busy, and its main job is to isolate "housekeeping" tasks (which aren't CPU bound) from the pCPUs that are used exclusively by vCPUs.

Well, this is how it is done today. If somebody doesn't want to allocate those extra cpus then they can go with the options we already have.

stefanha replied:
> I've been talking with @stefanha about this, and yes each iothread should be pin to a separate pCPU. Otherwise, just use a single iothread. @stefanha please correct me here if I'm wrong.

Yes.

alicefr commented:
> > It's also not entirely clear how a user would determine the right number of IO threads for a VM. ... In many cases, users may not need to configure these low-level details at all.
> 
> > The problem remains that KubeVirt doesn't know the best number and this is achieved by measuring the performance on the VM and setting that value from the experiments. So, this can be done only by the user
> 
> Thanks @alicefr for this PR and discussion!
> 
> I tend to agree with @vladikr's approach here. If we can't know how to configure this in the most optimized way, it would probably be very difficult for users to do so correctly.
> 
> In general, my opinion is that currently Kubevirt tends to give the admin/user a lot more control than he needs. This leads to a complicated API which demands a high expertise to define correctly. Our migration configurations are a good example for this, but there are a lot of other examples.
> 
> I would much rather use a declarative policy instead of a concrete number of threads, that will use a simple definition (i.e. always use 4 iothreads) or a heuristic (between 4 and 8 based on number of disks/vCPUs). Moving forward, the burden of optimizing this policy would be on Kubevirt, not on its users. It is possible that in the future we'd face a glass celling w.r.t. optimizing this and will see that users demand more control in order to fine tune, but let's not get ahead of ourselves and start simple. As @vladikr mentioned, going forward we might come up with tools that would be able to fine tune automatically an admin's intervention.

@iholder101 as already mentioned above, I don't think we can express this in a declarative policy. This needs to be a concrete number. Again, this number is an heuristics which should come from performance measurements done by the user. There is no magic formula based on number of disks/cpu which can set this number.

jeniferh commented:
> @jeniferh how would you distribute the iothreads? This PR pins each iothreads to the additional pCPU. Additionally, I have put as requested the `isolateEmulatorThread` because otherwise this will interfere either with the vcpu threads or the iothreads.
We can just let the host scheduler distribute the threads when dedicated/isolated pinning is not requested, I don't believe any pinning was used in that blog example where we see significant gains. On the K8s side I think the cpu request value can help in terms of not allowing too much overcommit, but my concern is users don't necessarily want to change a node level setting (i.e. enable CPUManager) just to test more IOthreads for a VM disk.

alicefr commented:
> > @jeniferh how would you distribute the iothreads? This PR pins each iothreads to the additional pCPU. Additionally, I have put as requested the `isolateEmulatorThread` because otherwise this will interfere either with the vcpu threads or the iothreads.
> 
> We can just let the host scheduler distribute the threads when dedicated/isolated pinning is not requested, I don't believe any pinning was used in that blog example where we see significant gains. On the K8s side I think the cpu request value can help in terms of not allowing too much overcommit, but my concern is users don't necessarily want to change a node level setting (i.e. enable CPUManager) just to test more IOthreads for a VM disk.

@stefanha I think we spoke about this offline. Can you please confirm that the iothreads should be scheduled on different pCPU, and therefore the pinning is necessary.

stefanha commented:
> Can you please confirm that the iothreads should be scheduled on different pCPU, and therefore the pinning is necessary.

Yes. IOThreads running on the same pCPU interfere with each other. They are intended to run on different CPUs for best performance. If not enough CPUs are available, then using fewer IOThreads would be more straightforward than having them compete on the same pCPU.

alicefr commented:
@stefanha @vladikr @stefanha after the recent discussions I have updated the PR and description. The main highlights:
1. Added a new field `policy` under the `IOThread` struct. This should replace the `iothreadPolicy` in this way we can specify a better semantic for the policy. In this case, the count for the number of iothreads
2. Added a new policy `dedicated`. If you have a better naming, please add your suggestions, I was short on ideas
3. Removed the restriction that dedicated cpus are mandatory. In case, the `dedicatedCpuPlacement` is false, then vcpu and iothreads will be scheduled by the host scheduler on all available cpus. Also in case of shared CPUs, the `dedicated` allocated additional CPUs equals to the number of `count`

vladikr commented:
> @stefanha @vladikr @stefanha after the recent discussions I have updated the PR and description. The main highlights:
> 
> 1. Added a new field `policy` under the `IOThread` struct. This should replace the `iothreadPolicy` in this way we can specify a better semantic for the policy. In this case, the count for the number of iothreads

On one hand, I like the the iothread.policy struct. Btw, shoun't it be IOthreads? 
On the other, I'm not sure what are we going to do with the current  `iothreadPolicy` filed, we can't just deprecate it.
What will happen if users with set both?

One way I could think of is to rename the  `IOThread` struct to something like  `IOThreadTune` and keep the `IOThreadPolicy`, while replacing all the existing policies with the `dedicated`. 
That way we will remove all the existing iothreads logic.
IOThreadTune can be optional struct and have the `count` (with a specific deafult)

> 2. Added a new policy `dedicated`. If you have a better naming, please add your suggestions, I was short on ideas

Works for me. Alternatively it could be `SharedPool`?

> 3. Removed the restriction that dedicated cpus are mandatory. In case, the `dedicatedCpuPlacement` is false, then vcpu and iothreads will be scheduled by the host scheduler on all available cpus. Also in case of shared CPUs, the `dedicated` allocated additional CPUs equals to the number of `count`

Based on @stefanha comment https://github.com/kubevirt/kubevirt/pull/13110#issuecomment-2460095003 how will we guarantee that iothreads are not running on the same pCPUs in the shared cpus case or is it best effort ?

alicefr commented:
> > @stefanha @vladikr @stefanha after the recent discussions I have updated the PR and description. The main highlights:
> > 
> > 1. Added a new field `policy` under the `IOThread` struct. This should replace the `iothreadPolicy` in this way we can specify a better semantic for the policy. In this case, the count for the number of iothreads
> 
> On one hand, I like the the iothread.policy struct. Btw, shoun't it be IOthreads? On the other, I'm not sure what are we going to do with the current `iothreadPolicy` filed, we can't just deprecate it. What will happen if users with set both?
> 

I was simply thinking of a warning message as we have for `running` vs the `runStrategy`. The new policy is available only with the new struct. Old VMs will continue working with the old policy type and field.

> One way I could think of is to rename the `IOThread` struct to something like `IOThreadTune` and keep the `IOThreadPolicy`, while replacing all the existing policies with the `dedicated`. That way we will remove all the existing iothreads logic. IOThreadTune can be optional struct and have the `count` (with a specific deafult)
> 

This would work, but I'm worry that will make the API even more chaotic. We already have the dedicatedCPU and isolatedEmulatorThread which has an influence on the feature. Do we want to take the chance to reorganize the API in a more clear way?

> > 2. Added a new policy `dedicated`. If you have a better naming, please add your suggestions, I was short on ideas
> 
> Works for me. Alternatively it could be `SharedPool`?
> 

I would prefer to avoid any word which is already in the other cpu policies to avoid any confusion. Even dedicated is too close to the dedicatedCPU for my taste. I would have choose some with `additional` or `extra`, but standalone they don't sound good either. 
How does `auxiliaryCPUs` or `supplementalCPUs` sound?

> > 3. Removed the restriction that dedicated cpus are mandatory. In case, the `dedicatedCpuPlacement` is false, then vcpu and iothreads will be scheduled by the host scheduler on all available cpus. Also in case of shared CPUs, the `dedicated` allocated additional CPUs equals to the number of `count`
> 
> Based on @stefanha comment [#13110 (comment)](https://github.com/kubevirt/kubevirt/pull/13110#issuecomment-2460095003) how will we guarantee that iothreads are not running on the same pCPUs in the shared cpus case or is it best effort ?
 
We won't, we will let vCPUs and IOThread float on the entire set of CPUs which is vCPU number + IOThreads count. In that mode, we simply add extra resources and rely on the host scheduler. Hence, performance aren't guaranteed as in the case of dedicated CPU

xpivarc commented:
> > > @stefanha @vladikr @stefanha after the recent discussions I have updated the PR and description. The main highlights:
> > > 
> > > 1. Added a new field `policy` under the `IOThread` struct. This should replace the `iothreadPolicy` in this way we can specify a better semantic for the policy. In this case, the count for the number of iothreads
> > 
> > 
> > On one hand, I like the the iothread.policy struct. Btw, shoun't it be IOthreads? On the other, I'm not sure what are we going to do with the current `iothreadPolicy` filed, we can't just deprecate it. What will happen if users with set both?
> 
> I was simply thinking of a warning message as we have for `running` vs the `runStrategy`. The new policy is available only with the new struct. Old VMs will continue working with the old policy type and field.
> 

The current IOThreadsPolicy should be sufficient for the goal outlined on the PR (My understanding is we just want an option to explicitly ask for number of IOthreads). The minimal change I see here is to either have the `count` field (or better a `IOThreadPool`) to be added. What we need to reason about is how does these additions interact with a) existing policies b) existing disk option `dedicatedIOThread` c) compatibility.


> > One way I could think of is to rename the `IOThread` struct to something like `IOThreadTune` and keep the `IOThreadPolicy`, while replacing all the existing policies with the `dedicated`. That way we will remove all the existing iothreads logic. IOThreadTune can be optional struct and have the `count` (with a specific deafult)
> 
> This would work, but I'm worry that will make the API even more chaotic. We already have the dedicatedCPU and isolatedEmulatorThread which has an influence on the feature. Do we want to take the chance to reorganize the API in a more clear way?

These are pretty contextual. The `DedicatedCPU` and `IsolatedEmulatorThread` do fall under `CPU` struct and they talk about placement of vCPUs to pCPUs. Here we should define a good context to distinguish this so it will be talking about IOThreads placement in  relation to pCPUs.

> 
> > > 2. Added a new policy `dedicated`. If you have a better naming, please add your suggestions, I was short on ideas
> > 
> > 
> > Works for me. Alternatively it could be `SharedPool`?
> 
> I would prefer to avoid any word which is already in the other cpu policies to avoid any confusion. Even dedicated is too close to the dedicatedCPU for my taste. I would have choose some with `additional` or `extra`, but standalone they don't sound good either. How does `auxiliaryCPUs` or `supplementalCPUs` sound?

I don't find this confusing as we give a good context `IOThreadsPolicy` cannot be miss-interpreted  by vCPU placement. 

> 
> > > 3. Removed the restriction that dedicated cpus are mandatory. In case, the `dedicatedCpuPlacement` is false, then vcpu and iothreads will be scheduled by the host scheduler on all available cpus. Also in case of shared CPUs, the `dedicated` allocated additional CPUs equals to the number of `count`
> > 
> > 
> > Based on @stefanha comment [#13110 (comment)](https://github.com/kubevirt/kubevirt/pull/13110#issuecomment-2460095003) how will we guarantee that iothreads are not running on the same pCPUs in the shared cpus case or is it best effort ?
> 
> We won't, we will let vCPUs and IOThread float on the entire set of CPUs which is vCPU number + IOThreads count. In that mode, we simply add extra resources and rely on the host scheduler. Hence, performance aren't guaranteed as in the case of dedicated CPU

Are we talking here about case where vCPU placement is not defined?

Generally I echo worries of Vladik, we are building a too level knobs and we don't build abstraction around this, we are risking that we bloat this just for us to come back and say we now recommend different settings and the API is not able to express it. There were good comments about Vladik and Itamar building this out of core API, be it policies or a general algorithm that will take hints and adjust the IOThreads accordingly. Such hints could be max IOps/disk(pvc), vCPUs,...

alicefr commented:
PTAL @vladikr @xpivarc

alicefr commented:
> > Again, there isn't such a formula. I would love to simplify the API and have some automatic way, but right now there isn't. The users needs to run some benchmarks and found out the best number for the iothreads count. This number isn't something that kubevirt can figure out.
> 
> But we did run some benchmarks and we should be able to deduce something from them or not? I am missing a lot of context from the benchmark we did. Like was these finding applied only for one disk?

@xpivarc you can read the findings in the blog post I referenced in the PR description.
Otherwise @stefanha @jeniferh are you able to give more details on this?

alicefr commented:
@xpivarc did we answers all the questions? 
Can you please take another look at the PR? Thanks!

alicefr commented:
/retest

vladikr commented:
From the API perspective, I think that this is an optimal approach considering the amount of information we have right now.
The API looks extensible enough for us to accommodate evolving scenarios. 

My only hesitation was about the `IOThreads` field name being too general. I was thinking in the direction of `IOThreadsPool` or `IOThreadsTune` or `IOThreadsControl`
I don't have a good name and this is pretty minor.

Thank you @alicefr !

/approve

alicefr commented:
> From the API perspective, I think that this is an optimal approach considering the amount of information we have right now. The API looks extensible enough for us to accommodate evolving scenarios.
> 
> My only hesitation was about the `IOThreads` field name being too general. I was thinking in the direction of `IOThreadsPool` or `IOThreadsTune` or `IOThreadsControl` I don't have a good name and this is pretty minor.
> 
I would personally prefer to keep it as generic as possible

alicefr commented:
Rebased and fixed a lint issue.

alicefr commented:
/test pull-kubevirt-manifests

akalenyu commented on staging/src/kubevirt.io/api/core/v1/schema.go:
```diff
@@ -1585,3 +1589,9 @@ type CPUTopology struct {
+type DiskIOThreads struct {
+	// Count specifies how many iothreads are allocated for the dedicated policy.
+	// +optional
+	Count *uint32 `json:"count,omitempty"`
```
Do we maybe want to be more explicit about the relationship between count and the supplementalpool policy? `supplementalPoolThreadCount`?
I am struggling to find a native k8s/k8s example but something along these lines:
https://github.com/openshift/api/blob/d6d8306a24ec146d7e8cb7bdc1259a2144dc77a0/config/v1alpha1/types_image_policy.go#L161-L164

alicefr replied:
Done, thanks!

akalenyu commented on pkg/virt-launcher/virtwrap/converter/converter.go:
```diff
@@ -1313,27 +1319,40 @@ func setIOThreads(vmi *v1.VirtualMachineInstance, domain *api.Domain, vcpus uint
 	ioThreadCount, autoThreads := getIOThreadsCountType(vmi)
 	if ioThreadCount != 0 {
 		if domain.Spec.IOThreads == nil {
 			domain.Spec.IOThreads = &api.IOThreads{}
```
random unrelated question. I could never figure out why kubevirt uses `<iothreads>` instead of the configurable `<iothreadids>`:
https://libvirt.org/formatdomain.html#iothreads-allocation

in this PRs context, is there any benefit in using the overloaded version?
I believe it's achievable without breaking the end kubevirt API user

alicefr replied:
Not sure, I get the question. But we are using both. The 'iothreads' defines how many iothreads in total we have, while each iothreadid allows a specific assignment for each disk. In this case, we are assigning all the iothreads to all the disks which require iothread, but you could have different configuration.

akalenyu replied:
okay just summarizing offline discussion, we don't use iothreadids in kubevirt and it's not relevant to this PR

akalenyu reviewed (commented):
Great to see such an impactful change go in with 0 func tests, just unit test coverage :pray: 

Some considerations, I will probably not insist on them but has to be brought up

alicefr commented on pkg/virt-controller/services/renderresources.go:
```diff
@@ -128,6 +145,16 @@ func WithoutDedicatedCPU(cpu *v1.CPU, cpuAllocationRatio int, withCPULimits bool
+func WithIOThreads(iothreads *v1.DiskIOThreads) ResourceRendererOption {
+	return func(renderer *ResourceRenderer) {
+		if iothreads == nil || iothreads.Count == nil || *iothreads.Count < 1 {
+			return
```
We never return an error in the With* () *Option functions

akalenyu replied:
gotcha

alicefr commented:
/test pull-kubevirt-e2e-k8s-1.32-sig-compute-serial
